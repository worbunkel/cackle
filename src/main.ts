import gql from 'graphql-tag';
import { graphql, GraphQLSchema, GraphQLObjectType, GraphQLString } from 'graphql';
import { print } from 'graphql/language/printer';
import _ from 'lodash';

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'RootQueryType',
    fields: {
      hello: {
        type: GraphQLString,
        resolve() {
          return 'world';
        },
      },
    },
  }),
});

const createDeepNames = (selections: any[]): any[] => {
  return _.flatMap(selections, selection => {
    const selectionSet = selection.selectionSet;
    if (!selectionSet) {
      return selection.name.value;
    }
    return _.map(createDeepNames(selectionSet.selections), name => selection.name.value + '.' + name);
  });
};

const printObjectQuery = (objectQuery: any): string => `{
  ${_.join(_.map(objectQuery, (value, key) => (value === '' ? key : `${key} ${printObjectQuery(value)}`)), '\n')}
}`;

let Requests: Array<{ AST: any; resolve: (val?: any) => void; reject: (val?: any) => void }> = [];

export const createQuery = async (query: any) => {
  new Promise((resolve, reject) => {
    Requests.push({
      AST: gql(query),
      resolve,
      reject,
    });
  });
};

const firstQuery = `
  query {
    hello
    hello2: hello
    test {
      thing
    }
  }
`;
const secondQuery = `
  {
    hello
    test2: test {
      thing
      thing2
      otherThing {
        moreTest
      }
    }
  }
`;

createQuery(firstQuery);
createQuery(secondQuery);

const processQueries = async () => {
  const queryDefinitionGroups = _.map(Requests, ({ AST }) => {
    const definitions = _.get(AST, 'definitions');
    // console.log(definitions);
    return _.filter(definitions, definition => _.get(definition, 'operation') === 'query');
  });
  // console.log(queryDefinitionGroups);
  const selections = _.map(queryDefinitionGroups, queryDefinitionGroup =>
    _.flatMap(queryDefinitionGroup, def => {
      return def.selectionSet.selections;
    }),
  );
  const deepNames = _.map(selections, createDeepNames);
  console.log('DEEP NAMES:\n', deepNames);
  const uniqueNames = _.uniq(...deepNames);
  const newQueryObject = {};
  _.each(uniqueNames, uniqueName => _.set(newQueryObject, uniqueName, ''));
  console.log(JSON.stringify(deepNames, null, 2));
  console.log(newQueryObject);
  const newQuery = printObjectQuery(newQueryObject);
  const finalAST = gql(newQuery);
  // const finalQueryDefinition = {
  //   kind: 'OperationDefinition',
  //   operation: 'query',
  //   variableDefinitions: [] as any[],
  //   directives: [] as any[],
  //   selectionSet: {
  //     kind: 'SelectionSet',
  //     selections: combinedSelections,
  //   },
  // };
  // const newAST = {
  //   kind: 'Document',
  //   definitions: [finalQueryDefinition],
  // };
  const query = print(finalAST as any);

  console.log('\nQUERY:\n', query);
  const result = await graphql(schema, query);
  console.log('RESULT:\n', JSON.stringify(result, null, 2));
  ASTs = [];
};

processQueries();
