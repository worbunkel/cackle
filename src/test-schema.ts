import { graphql, buildSchema, printSchema } from 'graphql';

var schema = buildSchema(`
type Mutation {
  addTodo(newTodo: NewTodoInput!): Todo!
  addTodos(newTodos: [NewTodoInput!]!): [Todo!]!
}

input NewTodoInput {
  name: String!
  isComplete: Boolean!
  user: NewUserInput
}

input NewUserInput {
  firstName: String!
  lastName: String!
}

type Query {
  todo(name: String!): Todo
  todos: [Todo!]!
}

type Todo {
  name: String!
  isComplete: Boolean!
  user: User!
}

type User {
  firstName: String!
  lastName: String!
  todos: [Todo!]!
}
`);

console.log(printSchema(schema));

type User = {
  firstName: string;
  lastName: string;
};

type Todo = {
  name: string;
  isComplete: boolean;
  user: User;
};

const getDefaultTodos = () => [
  {
    name: 'Brush Teeth',
    isComplete: true,
    user: {
      firstName: 'Test',
      lastName: 'User',
      todos: [
        {
          name: 'Brush Teeth',
          isComplete: true,
        },
      ],
    },
  },
  {
    name: 'Do Chores',
    isComplete: false,
    user: {
      firstName: 'Test',
      lastName: 'User2',
      todos: [
        {
          name: 'Do Chores',
          isComplete: false,
        },
      ],
    },
  },
];

class FakeDB {
  private todos: Todo[] = [];
  constructor() {
    this.todos = getDefaultTodos();
  }
  getTodos() {
    return this.todos;
  }
  getTodoByName(name: string) {
    return this.todos.find(todo => todo.name === name);
  }
  addTodo(newTodo: Todo) {
    this.todos.push(newTodo);
    return newTodo;
  }
  addTodos(newTodos: Todo[]) {
    this.todos = this.todos.concat(newTodos);
    return newTodos;
  }
  reset() {
    this.todos = getDefaultTodos();
  }
}

const fakeDB = new FakeDB();

var root = {
  todos: () => {
    return fakeDB.getTodos();
  },
  todo: ({ name }: { name: string }): Todo | undefined => {
    return fakeDB.getTodoByName(name);
  },
  addTodo: ({ newTodo }: { newTodo: Todo }) => {
    return fakeDB.addTodo(newTodo);
  },
  addTodos: ({ newTodos }: { newTodos: Todo[] }) => {
    return fakeDB.addTodos(newTodos);
  },
};

export const resetTestDB = () => {
  fakeDB.reset();
};

export const queryTestSchema = async (query: string) => {
  try {
    return await graphql(schema, query, root);
  } catch (err) {
    console.error(err);
  }
};
