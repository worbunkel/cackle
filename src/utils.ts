import _ from 'lodash';
import gql from 'graphql-tag';
import { print } from 'graphql/language/printer';

export type Request = {
  AST: any;
  resolve: (value?: any) => void;
  reject: (reason?: any) => void;
};

export const createQueryDefinitionGroupsFromASTs = (ASTs: any[]) =>
  _.map(ASTs, AST => {
    const definitions = _.get(AST, 'definitions');
    return _.filter(definitions, definition => _.get(definition, 'operation') === 'query');
  });

export const createMutationDefinitionGroupsFromASTs = (ASTS: any[]) =>
  _.map(ASTS, AST => {
    const definitions = _.get(AST, 'definitions');
    return _.filter(definitions, definition => _.get(definition, 'operation') === 'mutation');
  });

export const createSelections = (definitionGroups: any[][]) =>
  _.map(definitionGroups, definitionGroup =>
    _.flatMap(definitionGroup, definition => {
      return definition.selectionSet.selections;
    }),
  );

const createArgumentsName = (selection: any) => {
  if (_.size(selection.arguments) > 0) {
    return `(${_.map(
      selection.arguments,
      argument =>
        `${argument.name.value}:${
          argument.value.kind === 'IntValue' ? argument.value.value : JSON.stringify(argument.value.value)
        }`,
    )})`;
  }
  return '';
};

const createDeepNames = (selections: any[]): any[] => {
  return _.flatMap(selections, selection => {
    const selectionSet = selection.selectionSet;
    if (!selectionSet) {
      return selection.name.value;
    }
    const argumentsName = createArgumentsName(selection);
    const nameWithArguments = argumentsName
      ? `${selection.name.value}${_.replace(argumentsName, /[^a-zA-Z0-9]/g, '')}: ${
          selection.name.value
        }${argumentsName}`
      : selection.name.value;
    return _.map(createDeepNames(selectionSet.selections), name => nameWithArguments + '.' + name);
  });
};

const createDeepAliases = (selections: any[]): any[] => {
  return _.flatMap(selections, selection => {
    const selectionSet = selection.selectionSet;
    const alias = _.get(selection, 'alias.value', selection.name.value);
    if (!selectionSet) {
      return alias;
    }
    return _.map(createDeepAliases(selectionSet.selections), childAlias => alias + '.' + childAlias);
  });
};

// TODO: Clean this up to be done in one loop
export const createDeepNamesAndAliases = (selections: any[]): { deepNames: any[]; deepAliases: any[] } => ({
  deepNames: _.map(selections, selection => createDeepNames(selection)),
  deepAliases: _.map(selections, selection => createDeepAliases(selection)),
});

export const printObjectQuery = (objectQuery: any): string => `{
  ${_.join(_.map(objectQuery, (value, key) => (value === '' ? key : `${key} ${printObjectQuery(value)}`)), '\n')}
}`;

export const createUniqueNames = (deepNames: string[]) => _.uniq(_.flatten(deepNames));

export const createQueryFromUniqueNames = (uniqueNames: any[]) => {
  const newQueryObject = {};
  _.each(uniqueNames, uniqueName => _.set(newQueryObject, uniqueName, ''));
  const newQuery = printObjectQuery(newQueryObject);
  const finalAST = gql(newQuery);
  const query = print(finalAST as any);
  return query;
};

export const createQueryNamesAndAliasesFromASTs = (ASTs: any[]) => {
  const queryDefinitionGroups = createQueryDefinitionGroupsFromASTs(ASTs);
  const selections = createSelections(queryDefinitionGroups);
  const { deepNames, deepAliases } = createDeepNamesAndAliases(selections);
  const uniqueNames = createUniqueNames(deepNames);
  const query = createQueryFromUniqueNames(uniqueNames);
  return { query, deepNames, deepAliases };
};

const createMutationSelectionTitleGroups = (selectionGroups: any[][]) => {
  const selectionTitles: string[] = [];
  return _.map(selectionGroups, selectionGroup =>
    _.map(selectionGroup, selection => {
      const selectionAlias = _.get(selection, 'alias.value');
      const selectionName = _.get(selection, 'name.value');
      const selectionTitle = selectionAlias || selectionName || '';
      const numMatchingTitles = _.sumBy(selectionTitles, existingSelectionTitle =>
        existingSelectionTitle === selectionTitle ? 1 : 0,
      );
      selectionTitles.push(selectionTitle);
      const needsAlias = numMatchingTitles > 0;
      const newSelectionTitle = numMatchingTitles > 0 ? `${selectionTitle}${numMatchingTitles + 1}` : selectionTitle;
      if (needsAlias) {
        _.set(selection, 'alias', {
          kind: 'Name',
          value: newSelectionTitle,
        });
      }
      return { title: { original: selectionTitle, new: newSelectionTitle }, mutationString: print(selection) };
    }),
  );
};

export const createMutationAndNamesFromASTs = (ASTs: any[]) => {
  const mutationDefinitionGroups = createMutationDefinitionGroupsFromASTs(ASTs);
  const selectionGroups = createSelections(mutationDefinitionGroups);
  const selectionTitleGroups = createMutationSelectionTitleGroups(selectionGroups);
  const newMutation = print(
    gql(`mutation{
    ${_.map(_.flatten(selectionTitleGroups), selectionTitleGroup => selectionTitleGroup.mutationString).join('\n')}
  }`),
  );
  const names = _.map(selectionTitleGroups, selectionTitleGroup =>
    _.map(selectionTitleGroup, selectionTitle => selectionTitle.title),
  );
  return {
    mutation: newMutation,
    names,
  };
};

export const delay = async (durationInMS: number) =>
  new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, durationInMS);
  });
