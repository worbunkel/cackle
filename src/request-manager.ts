import gql from 'graphql-tag';
import _ from 'lodash';
import { graphql } from 'graphql';
import { print } from 'graphql/language/printer';
import { createDeepNamesAndAliases, printObjectQuery } from './utils';
import { testSchema } from './test-schema';

type Request = {
  AST: any;
  resolve: (value?: any) => void;
  reject: (val?: any) => void;
};

export class RequestManager {
  requests: Array<Request> = [];
  constructor() {
    this.requests = [];
  }
  async createQuery(query: any) {
    return new Promise((resolve, reject) => {
      this.requests.push({
        AST: gql(query),
        resolve,
        reject,
      });
    });
  }
  async processQueries() {
    const queryDefinitionGroups = _.map(this.requests, ({ AST }) => {
      const definitions = _.get(AST, 'definitions');
      return _.filter(definitions, definition => _.get(definition, 'operation') === 'query');
    });
    console.log('queryDefinitionGroups');
    console.log(queryDefinitionGroups);
    const selections = _.map(queryDefinitionGroups, queryDefinitionGroup =>
      _.flatMap(queryDefinitionGroup, def => {
        return def.selectionSet.selections;
      }),
    );
    console.log(selections);
    const { deepNames, deepAliases } = createDeepNamesAndAliases(selections);
    console.log('DEEP NAMES:\n', deepNames);
    console.log('DEEP ALIASES:\n', deepAliases);
    const uniqueNames = _.uniq(_.flatten(deepNames));
    console.log('Unique Names:\n', uniqueNames);
    const newQueryObject = {};
    _.each(uniqueNames, uniqueName => _.set(newQueryObject, uniqueName, ''));
    console.log(newQueryObject);
    const newQuery = printObjectQuery(newQueryObject);
    const finalAST = gql(newQuery);
    const query = print(finalAST as any);

    console.log('\nQUERY:\n', query);

    const response = await graphql(testSchema, query);
    const result = response.data;
    console.log(JSON.stringify(result, null, 2));

    _.each(deepNames, (deepNameGroup, index) => {
      const deepAliasGroup = deepAliases[index];
      const returnObj = {};
      _.each(deepAliasGroup, (deepAlias, aliasIndex) => {
        const deepName = deepNameGroup[aliasIndex];
        _.set(returnObj, deepAlias, _.get(result, deepName));
      });
      this.requests[index].resolve(returnObj);
    });

    this.requests = [];
  }
}
