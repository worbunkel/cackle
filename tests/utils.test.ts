import {
  createQueryFromUniqueNames,
  createQueryDefinitionGroups,
  createSelections,
  createDeepNamesAndAliases,
  delay,
} from '../dist/utils';
import gql from 'graphql-tag';
import _ from 'lodash';
import { RequestManager } from '../src/request-manager';
import { graphql } from 'graphql';
import { testSchema } from '../src/test-schema';

const firstQuery = `
  query {
    hello
    hello2: hello
    test {
      thing
    }
  }
`;

describe('utils', () => {
  describe('createDeepNamesAndAliases', () => {
    it('should create deepNames and deepAliases correctly', () => {
      const requests = [
        {
          AST: gql(firstQuery),
          resolve: (value: any) => {},
          reject: (reason: any) => {},
        },
      ];
      const queryDefinitionGroups = createQueryDefinitionGroups(requests);
      const selections = createSelections(queryDefinitionGroups);

      const deepNamesAndAliases = createDeepNamesAndAliases(selections);

      expect(deepNamesAndAliases).toEqual({
        deepNames: [['hello', 'hello', 'test.thing']],
        deepAliases: [['hello', 'hello2', 'test.thing']],
      });
    });
  });

  describe('createQueryFromUniqueNames', () => {
    it('should combine unique names into a query', () => {
      const uniqueNames = ['hello', 'test.thing', 'test.thing2', 'test.otherThing.moreTest'];

      const query = createQueryFromUniqueNames(uniqueNames);

      const expectedResult = `{
  hello
  test {
    thing
    thing2
    otherThing {
      moreTest
    }
  }
}
`;
      expect(query).toEqual(expectedResult);
    });
  });

  describe('createQueryNamesAndAliasesFromASTs', () => {
    const queryWithOneArgument = `
    {
      thingWithArgs(argument1: "test"){
        thingInside
      }
    }
    `;

    const queryWithOneArgument2 = `
    {
      thingWithArgs(argument1: "test2"){
        thingInside
      }
    }
    `;

    const queryWithOneArgumentAndAlias = `
    {
      thingWithArgs1: thingWithArgs(argument1: "test"){
        otherThingInside
      }
    }`;

    const queryWithTwoArguments = `
    {
      thingWithArgs(argument1: "noAlias", argument2: "testAgain"){
        thingInside
      }
    }
    `;

    const queryWithTwoArguments2 = `
    {
      thingWithArgs(argument1: "test2", argument2: "testAgain"){
        thingInside
      }
    }
    `;

    const queriesWithOneArgumentMatched = [queryWithOneArgument, queryWithOneArgument];
    const queriesWithOneArgumentUnmatched = [queryWithOneArgument, queryWithOneArgument2];
    const queriesWithOneArgumentMatchedAndAlias = [queryWithOneArgument, queryWithOneArgumentAndAlias];
    const queriesWithTwoArgumentsMatched = [queryWithTwoArguments, queryWithTwoArguments];
    const queriesWithTwoArgumentsUnmatched = [queryWithTwoArguments, queryWithTwoArguments2];
    const queriesWithAllVariations = [
      queryWithOneArgument,
      queryWithOneArgument2,
      queryWithOneArgumentAndAlias,
      queryWithTwoArguments,
      queryWithTwoArguments2,
    ];
    const getRequestQuery = async (queries: string[]) => {
      let requestQuery = '';
      const queryChecker = async (query: string) => {
        requestQuery = query;
        return { data: {} };
      };
      const requestManager = new RequestManager(queryChecker);
      const resultPromises = _.map(queries, query => requestManager.createQuery(query));
      await Promise.all(resultPromises);
      return requestQuery;
    };
    it('Can handle queries one argument matched', async () => {
      expect(await getRequestQuery(queriesWithOneArgumentMatched)).toMatchSnapshot();
    });
    it('Can handle queries one argument unmatched', async () => {
      expect(await getRequestQuery(queriesWithOneArgumentUnmatched)).toMatchSnapshot();
    });
    it('Can handle queries with one argument matched and alias', async () => {
      expect(await getRequestQuery(queriesWithOneArgumentMatchedAndAlias)).toMatchSnapshot();
    });
    it('Can handle queries with two arguments matched', async () => {
      expect(await getRequestQuery(queriesWithTwoArgumentsMatched)).toMatchSnapshot();
    });
    it('Can handle queries with two arguments unmatched', async () => {
      expect(await getRequestQuery(queriesWithTwoArgumentsUnmatched)).toMatchSnapshot();
    });
    it('Can handle queries with many variations', async () => {
      expect(await getRequestQuery(queriesWithAllVariations)).toMatchSnapshot();
    });
  });

  describe('delay', () => {
    const durations = [100, 200, 300, 400, 500, 1000];
    _.each(durations, duration => {
      it(`should allow at least ${duration}ms to pass when passed ${duration} as the duration`, async () => {
        const timeBefore = new Date().getTime();
        await delay(duration);
        const timeAfter = new Date().getTime();

        const difference = timeAfter - timeBefore;
        expect(difference + 1).toBeGreaterThan(duration);
      });
    });
  });
});
