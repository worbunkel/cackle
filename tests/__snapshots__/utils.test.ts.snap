// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils createQueryNamesAndAliasesFromASTs Can handle queries one argument matched 1`] = `
"{
  thingWithArgsargument1test: thingWithArgs(argument1: \\"test\\") {
    thingInside
  }
}
"
`;

exports[`utils createQueryNamesAndAliasesFromASTs Can handle queries one argument unmatched 1`] = `
"{
  thingWithArgsargument1test: thingWithArgs(argument1: \\"test\\") {
    thingInside
  }
  thingWithArgsargument1test2: thingWithArgs(argument1: \\"test2\\") {
    thingInside
  }
}
"
`;

exports[`utils createQueryNamesAndAliasesFromASTs Can handle queries with many variations 1`] = `
"{
  thingWithArgsargument1test: thingWithArgs(argument1: \\"test\\") {
    thingInside
    otherThingInside
  }
  thingWithArgsargument1test2: thingWithArgs(argument1: \\"test2\\") {
    thingInside
  }
  thingWithArgsargument1noAliasargument2testAgain: thingWithArgs(argument1: \\"noAlias\\", argument2: \\"testAgain\\") {
    thingInside
  }
  thingWithArgsargument1test2argument2testAgain: thingWithArgs(argument1: \\"test2\\", argument2: \\"testAgain\\") {
    thingInside
  }
}
"
`;

exports[`utils createQueryNamesAndAliasesFromASTs Can handle queries with number arguments 1`] = `
"{
  thingWithArgsargument11: thingWithArgs(argument1: 1) {
    thingInside
  }
}
"
`;

exports[`utils createQueryNamesAndAliasesFromASTs Can handle queries with one argument matched and alias 1`] = `
"{
  thingWithArgsargument1test: thingWithArgs(argument1: \\"test\\") {
    thingInside
    otherThingInside
  }
}
"
`;

exports[`utils createQueryNamesAndAliasesFromASTs Can handle queries with two arguments matched 1`] = `
"{
  thingWithArgsargument1noAliasargument2testAgain: thingWithArgs(argument1: \\"noAlias\\", argument2: \\"testAgain\\") {
    thingInside
  }
}
"
`;

exports[`utils createQueryNamesAndAliasesFromASTs Can handle queries with two arguments unmatched 1`] = `
"{
  thingWithArgsargument1noAliasargument2testAgain: thingWithArgs(argument1: \\"noAlias\\", argument2: \\"testAgain\\") {
    thingInside
  }
  thingWithArgsargument1test2argument2testAgain: thingWithArgs(argument1: \\"test2\\", argument2: \\"testAgain\\") {
    thingInside
  }
}
"
`;
